8365633: JFR reports incorrect number of cores on hybrid CPU

YaSuenag opened the pull request:
On hybrid CPU like Intel Arrow Lake, JFR reports incorrect number of cores in `jdk.CPUInformation`.

```
$ jfr print --events jdk.CPUInformation test.jfr
jdk.CPUInformation {
  startTime = 10:49:27.692 (2025-09-04)
  cpu = "Intel (null) (HT) SSE SSE2 SSE3 SSSE3 SSE4.1 SSE4.2 Core Intel64"
  description = "Brand: Intel(R) Core(TM) Ultra 5 225U, Vendor: GenuineIntel
Family: <unknown> (0x6), Model: <unknown> (0xb5), Stepping: 0x0
Ext. family: 0x0, Ext. model: 0xb, Type: 0x0, Signature: 0x000b0650
Features: ebx: 0x40800800, ecx: 0xfffaf38b, edx: 0xbfcbfbff
Ext. features: eax: 0x00000000, ebx: 0x00000000, ecx: 0x00000121, edx: 0x2c100800
Supports: On-Chip FPU, Virtual Mode Extensions, Debugging Extensions, Page Size Extensions, Time Stamp Counter, Model Specific Registers, Physical Address Extension, Machine Check Exceptions, CMPXCHG8B Instruction, On-Chip APIC, Fast System Call, Memory Type Range Registers, Page Global Enable, Machine Check Architecture, Conditional Mov Instruction, Page Attribute Table, 36-bit Page Size Extension, CLFLUSH Instruction, ACPI registers in MSR space, Intel Architecture MMX Technology, Fast Float Point Save and Restore, Streaming SIMD extensions, Streaming SIMD extensions 2, Self-Snoop, Hyper Threading, Thermal Monitor, Streaming SIMD Extensions 3, PCLMULQDQ, MONITOR/MWAIT instructions, Enhanced Intel SpeedStep technology, Thermal Monitor 2, Supplemental Streaming SIMD Extensions 3, Fused Multiply-Add, CMPXCHG16B, xTPR Update Control, Perfmon and Debug Capability, Process-context identifiers, Streaming SIMD extensions 4.1, Streaming SIMD extensions 4.2, x2APIC, MOVBE, Popcount instruction, TSC-Deadline, AESNI, XSAVE, OSXSAVE, AVX, F16C, LAHF/SAHF instruction support, Advanced Bit Manipulations: LZCNT, SYSCALL/SYSRET, Execute Disable Bit, RDTSCP, Intel 64 Architecture, Invariant TSC"
  sockets = 1
  cores = 7
  hwThreads = 14
}
```

Flight record in above was created on Intel Core Ultra 5 225U. According to [datasheet](https://www.intel.com/content/www/us/en/products/sku/241861/intel-core-ultra-5-processor-225u-12m-cache-up-to-4-80-ghz/specifications.html), it has 12 cores, 14 threads. Thus JFR should report `12` in `cores`.

We tackled similar issue on [JDK-8365633](https://bugs.openjdk.org/browse/JDK-8365633), then we concluded it is difficult to calculate number of physical cores. Thus we might not set correct value at here, but we should avoid to set incorrect value at least.

This patch sets `0` to `cores` and "Hybrid Architecture" is added to `description` as following if `jdk.CPUInformation` if it is generated on hybrid CPU. I want to set `-1` to `cores` TBH, but I didn't because `cores` is declared as `uint` - `-1` is equivalent with `UINT_MAX`, it looks like strange at here.

```
jdk.CPUInformation {
  startTime = 10:48:12.420 (2025-09-04)
  cpu = "Intel (null) (HT) SSE SSE2 SSE3 SSSE3 SSE4.1 SSE4.2 Core Intel64"
  description = "Brand: Intel(R) Core(TM) Ultra 5 225U, Vendor: GenuineIntel
Family: <unknown> (0x6), Model: <unknown> (0xb5), Stepping: 0x0
Ext. family: 0x0, Ext. model: 0xb, Type: 0x0, Signature: 0x000b0650
Features: ebx: 0x42800800, ecx: 0xfffaf38b, edx: 0xbfcbfbff
Ext. features: eax: 0x00000000, ebx: 0x00000000, ecx: 0x00000121, edx: 0x2c100800
Supports: On-Chip FPU, Virtual Mode Extensions, Debugging Extensions, Page Size Extensions, Time Stamp Counter, Model Specific Registers, Physical Address Extension, Machine Check Exceptions, CMPXCHG8B Instruction, On-Chip APIC, Fast System Call, Memory Type Range Registers, Page Global Enable, Machine Check Architecture, Conditional Mov Instruction, Page Attribute Table, 36-bit Page Size Extension, CLFLUSH Instruction, ACPI registers in MSR space, Intel Architecture MMX Technology, Fast Float Point Save and Restore, Streaming SIMD extensions, Streaming SIMD extensions 2, Self-Snoop, Hyper Threading, Thermal Monitor, Streaming SIMD Extensions 3, PCLMULQDQ, MONITOR/MWAIT instructions, Enhanced Intel SpeedStep technology, Thermal Monitor 2, Supplemental Streaming SIMD Extensions 3, Fused Multiply-Add, CMPXCHG16B, xTPR Update Control, Perfmon and Debug Capability, Process-context identifiers, Streaming SIMD extensions 4.1, Streaming SIMD extensions 4.2, x2APIC, MOVBE, Popcount instruction, TSC-Deadline, AESNI, XSAVE, OSXSAVE, AVX, F16C, LAHF/SAHF instruction support, Advanced Bit Manipulations: LZCNT, SYSCALL/SYSRET, Execute Disable Bit, RDTSCP, Intel 64 Architecture, Invariant TSC, Hybrid Architecture"
  sockets = 1
  cores = 0
  hwThreads = 14
}
```

This change affects showing information only, would not change the behavior like active processor count.

YaSuenag commented:
/label hotspot-jfr

YaSuenag commented:
/issue JDK-8366847

YaSuenag commented:
/issue remove JDK-8365633

YaSuenag commented on src/hotspot/cpu/x86/vm_version_x86.cpp:
```diff
@@ -2573,7 +2573,12 @@ void VM_Version::resolve_cpu_information_details(void) {
 
   // find out number of threads per cpu package
-  int threads_per_package = threads_per_core() * cores_per_cpu();
+  int threads_per_package = _cpuid_info.tpl_cpuidB1_ebx.bits.logical_cpus;
+  if (threads_per_package == 0) {
+    // Fallback code to avoid div by zero in subsequent code.
+    // CPUID 0Bh (ECX = 1) might return 0 on older AMD processor (EPYC 7763 at least)
+    threads_per_package = threads_per_core() * cores_per_cpu();
+  }
```
Check `thread_per_package` from `CPUID` to avoid SIGFPE (div by zero).

`CPUID` leaf 0Bh with ECX (subleaf) = 1 seems to return `0` as number of logical processors in spite of leaf 08h is supported in some older processors. It's strange, but I saw it on AMD EPYC 7763 on GitHub Actions runner. The problem appears as following.

I haven't faced this problem on AMD Ryzen 3 3300X.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGFPE (0x8) at pc=0x00007f19e11ffeed, pid=11344, tid=11345
#
# JRE version: OpenJDK Runtime Environment (26.0) (build 26-internal-YaSuenag-ebf4aec23fa95c8d54d1196c7be85e99f0846420)
# Java VM: OpenJDK 64-Bit Server VM (26-internal-YaSuenag-ebf4aec23fa95c8d54d1196c7be85e99f0846420, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x11ffeed]  VM_Version::initialize_cpu_information()+0x2d
#
# CreateCoredumpOnCrash turned off, no core file dumped
#
# JFR recording file will be written. Location: /home/runner/work/jdk/jdk/build/run-test-prebuilt/test-support/jtreg_test_jdk_tier1_part3/scratch/1/hs_err_pid11344.jfr
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#

---------------  S U M M A R Y ------------

Command Line: -Xmx768m -XX:MaxRAMPercentage=12.5 -Dtest.boot.jdk=/home/runner/work/jdk/jdk/bootjdk/jdk -Djava.io.tmpdir=/home/runner/work/jdk/jdk/build/run-test-prebuilt/test-support/jtreg_test_jdk_tier1_part3/tmp -ea -esa -XX:-CreateCoredumpOnCrash -Xlog:jfr=trace -XX:StartFlightRecording:filename=./dumped.jfr,dumponexit=true,settings=profile jdk.jfr.startupargs.TestDumpOnExit$TestMain

Host: AMD EPYC 7763 64-Core Processor, 4 cores, 15G, Ubuntu 22.04.5 LTS
Time: Thu Sep  4 07:17:02 2025 UTC elapsed time: 0.570221 seconds (0d 0h 0m 0s)

---------------  T H R E A D  ---------------

Current thread (0x00007f19d802ad20):  JavaThread "main"             [_thread_in_vm, id=11345, stack(0x00007f19dff00000,0x00007f19e0000000) (1024K)]

Stack: [0x00007f19dff00000,0x00007f19e0000000],  sp=0x00007f19dfffbd40,  free space=1007k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x11ffeed]  VM_Version::initialize_cpu_information()+0x2d  (vm_version_x86.cpp:2581)
```

dholmes-ora replied:
Sorry I don't follow this change. If there is a division-by-zero problem then don't we need a fix the method that would report zero i.e. `threads_per_core()` or `cores_per_cpu()` ??

YaSuenag replied:
I want to use number of logical processors from `CPUID` instruction directly because we can't believe `threads_per_core()` on hybrid CPU. Some of cores (e.g. E cores) might not have hyper threading even though `CPUID` reports HT flag.
Thus I think we have to check the value here.

YaSuenag replied:
> Okay but this is not a JFR specific change - you are changing these values for all clients of the VM version info.

Yes, so I think this PR needs reviewer(s) from HotSpot folks.

> So the current code would do:

Unfortunately, no. It would derive wrong number of sockets as following:

```
_no_of_threads = 28  // 2 sockets * (6 P-cores * 2 threads + 2 E-cores)
threads_per_core()  = 2   // HT enabled
cores_per_cpu() = 7 // ** different from your thought ** 14 threads per cpu / 2 logical processors
threads_per_package = 14 // 2 threads per core * 7 cores per cpu
_no_of_sockets = 2 // 28 threads / 14 threads per package
```

`cores_per_cpu()` would return `CPUID` leaf 0Bh for modern Intel CPU as following. According to Software Developer's Manual, leaf 0Bh returns number of "logical" processor in each domains (selected by subleaf (ECX)).

```cpp
  if (is_intel()) {
    bool supports_topology = supports_processor_topology();
    if (supports_topology) {
      result = _cpuid_info.tpl_cpuidB1_ebx.bits.logical_cpus /
               _cpuid_info.tpl_cpuidB0_ebx.bits.logical_cpus;
    }
```

In this PR, all of variables are same with current implementation, thus we can calculate `_no_of_sockets` correctly. But only `threads_per_package` comes from `CPUID`.

```
_no_of_threads = 28  // 2 sockets * (6 P-cores * 2 threads + 2 E-cores)
threads_per_core()  = 2   // HT enabled
cores_per_cpu() = 7 // incorrect, but it would be calculated by values from `CPUID`
threads_per_package = 14 // ** different from current implementation ** the value from `CPUID`
_no_of_sockets = 2 // 28 threads / 14 threads per package
```

I uploaded test code for `CPUID` leaf  0Bh: https://github.com/YaSuenag/garakuta/blob/master/check-hybrid-cores/hy-core-check.c

YaSuenag replied:
Agree! HT flag is set even though it is E-core (do not have HT), and also we cannot get number of physical cores, hence cores-per-cpu is always incorrect as you said.

dholmes-ora replied:
Can we not check for "Intel Hybrid Information" CPUID leaf (0x1A) that reports the type of each core and use that to get an accurate core count? So that:
```
int cores_per_cpu() {
  if (is_hybrid()) {
    if (hyper_threaded() ) {
      return _no_of_threads - _no_of_e_cores;
     } else {
      return _no_of_threads;
     }
  } else {
    return _no_of_threads / threads_per_core();
  }
}
```

YaSuenag replied:
I think it is difficult for the following reasons:

1. We need to set affinity & issue `CPUID` leaf 1Ah on all of cores. See example: https://github.com/YaSuenag/garakuta/blob/master/check-hybrid-cores/hy-core-check.c
2. E-core does not have HT for now, but we don't know this architecture (P core has HT, both E core and LP E core do not have HT) keeps in future.

dholmes-ora replied:
Oh that is awful - I did not realize that, I assumed we got some kind of bit vector we could query.

I'm less concerned about them adding HT to E-cores in the future as we will likely have to expand the code to deal with specific chips anyway.

So where does this leave us?
- `cores_per_cpu` is an approximation
-  anything reporting the `cores_per_cpu` value needs to include some text (if it is a hybrid) to indicate it is an approximation

dcubed-ojdk replied:
Based on the comments that I'm reading in this PR, it seems like the existing logic
presumes Symmetric Multi-Processing (SMP) instead of Asymmetric Multi-Processing
(AMP) or is the AMP support in a different place in the code.

A very long time ago, Solaris SPARC servers could be provisioned with differing processors
with different core counts and speed ratings. I don't know if we supported those AMP
servers in Java or not, but there was definitely support in Solaris itself.

YaSuenag replied:
I updated this PR to revert data type change to `uint`, so number of cores might be incorrect on hybrid CPU. "cores" in `jdk.CPUInformation` event would be set "approximated" value, but "Hybrid Architecture" would be added into "description" field in that case.

I grep'ed to find user of `cores_per_cpu`, this value would be used in JFR and reporting only like I've fixed in #26808 , thus this change affects limited area only AFAICS.

> Based on the comments that I'm reading in this PR, it seems like the existing logic
presumes Symmetric Multi-Processing (SMP) instead of Asymmetric Multi-Processing
(AMP) or is the AMP support in a different place in the code.

Has Intel released AMP processor for Xeon / Core line? If AMP is SPARC only in context of Java, it would be implemented on SPARC-specific code, but it has already gone...

dholmes-ora replied:
@dcubed-ojdk my recollection is that the "symmetry" in SMP related to memory accesses - but there are differing definitions. There is also a presumption that all "logical processors" behave effectively the same, and for multi-socket systems all CPU's are the same. That is still true-enough with these Hybrid chips, they just don't have the same internal topology so we can't coarsely assume we can calculate things like "number of cores" by knowing number-of-threads and whether hyper-threading is enabled. Unfortunately you also cannot easily get the exact details of that topology.

YaSuenag commented:
> If a value is missing, emit jmc_undefined_long in event.set_cores(...), so it will show up as N/A in both JMC and the JFR tool.

It couldn't for now because `cores` is declared as `uint`.
```xml
<Field type="uint" name="cores" label="Cores" />
```
I checked generated code from `metadata.xml`, the field as declared as `uint` (32 bit), so we cannot store `jmc_undefined_long` as `long`.
```cpp
class EventCPUInformation : public JfrEvent<EventCPUInformation>
{
 private:
  const char* _cpu;
  const char* _description;
  unsigned _sockets;
  unsigned _cores;
  unsigned _hwThreads;
```

But I agree with @egahlin to show up "N/A" in this case. I think it is the most simple way to change data type to `long` for number of CPU cores in `metadata.xml`. Can we do that? I concern the impact for JFR recording data format.

As an another option, we can update `EventPrintWriter::getUnsigned` to handle `unsigned int` - it might be reasonable to fix this problem.

Which solution is better?

YaSuenag commented:
@egahlin 

`event.getInt("cores")`: failed
```
java.lang.IllegalArgumentException: Attempt to get field "cores" with illegal data type conversion int
        at jdk.jfr/jdk.jfr.consumer.RecordedObject.newIllegalArgumentException(RecordedObject.java:890)
        at jdk.jfr/jdk.jfr.consumer.RecordedObject.getInt(RecordedObject.java:553)
        at jdk.jfr.event.os.TestCPUInformation.main(TestCPUInformation.java:56)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
        at java.base/java.lang.Thread.run(Thread.java:1474)
```

`Integer value = event.getValue("cores")`: failed
```
java.lang.ClassCastException: class java.lang.Long cannot be cast to class java.lang.Integer (java.lang.Long and java.lang.Integer are in module java.base of loader 'bootstrap')
        at jdk.jfr.event.os.TestCPUInformation.main(TestCPUInformation.java:57)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
        at java.base/java.lang.Thread.run(Thread.java:1474)
```

YaSuenag commented:
@egahlin Could you review?

YaSuenag commented:
Can someone review this PR? I think I have to got reviewers both HotSpot and JFR.

YaSuenag commented:
I saw similar behavior in `VM.info` dcmd, and I modified the output for CPU information in #26808 - removed thread-related information, and added "hybrid" flag.

But I think we cannot apply similar method in JFR because we have to treat number of cores even if "hybrid" flag would be added in event definition, and also we might change all of viewer (JMC, jfr command, and more) to hide number of cores if hybrid flag is set. It is impractical I think.

So I think it is the most reasonable to set `min_jlong` which already means "N/A".

egahlin commented:
> So I think it is the most reasonable to set `min_jlong` which already means "N/A".

I don't think using the min_jlong value is the problem, we do that for other event values, but promoting the field to a long value might be too disruptive. It's a value users are likely to extract programmatically.

YaSuenag commented:
> but promoting the field to a long value might be too disruptive. It's a value users are likely to extract programmatically.

If the data size is important than showing "N/A", I think we can set "0" for number of cores as I proposed [at first](https://openjdk.github.io/cr/?repo=jdk&pr=27080&range=00). But you said before that it should not be a problem to change to type long: https://github.com/openjdk/jdk/pull/27080#issuecomment-3270024207

Which is better? I think it is better not to change data size rather than showing "N/A" for compatibility. I think it is whether it feels natural for JFR that "0" means "N/A".

egahlin commented:
>  But you said before that it should not be a problem to change to type long: [#27080 (comment)](https://github.com/openjdk/jdk/pull/27080#issuecomment-3270024207)

I looked at it from a file format perspective (varint encoding), but from an API perspective, as you showed, both getInt("cores") and getValue("cores") may throw an exception. This could impact users who have no problem with hybrid cores today. Also, their code may work on a development or test machine but then break on production hardware.

jcmd is meant for humans to read, so a change there has less impact.

YaSuenag commented:
So should I revert the change to original? I think it is better not to change data type. Compatibility should prefer than user friendly message ("N/A") in this case.

YaSuenag commented on src/hotspot/cpu/x86/vm_version_x86.cpp:
```diff
@@ -2585,8 +2590,9 @@ void VM_Version::resolve_cpu_information_details(void) {
 
-  // estimate the number of cores
-  _no_of_cores = cores_per_cpu() * _no_of_sockets;
+  // estimate the number of cores.
+  // -1 if hybrid CPU because it is difficult to derive number of cores.
+  _no_of_cores = supports_hybrid() ? -1 : (cores_per_cpu() * _no_of_sockets);
```
I think incorrect value is not useful because it causes confusing.

If we have to keep current code to set estimated value, I can use `VM_Version::supports_hybrid()` in JFR code to emit -1 on the event.

egahlin replied:
Even if the value is incorrect, it still has some value as an approximation. We could update the description to make this clearer, or perhaps provide additional information to help users.

dholmes-ora commented on src/hotspot/cpu/x86/vm_version_x86.cpp:
```diff
@@ -2573,7 +2573,12 @@ void VM_Version::resolve_cpu_information_details(void) {
+  int threads_per_package = _cpuid_info.tpl_cpuidB1_ebx.bits.logical_cpus;
+  if (threads_per_package == 0) {
+    // Fallback code to avoid div by zero in subsequent code.
+    // CPUID 0Bh (ECX = 1) might return 0 on older AMD processor (EPYC 7763 at least)
```
Can we assert that this is not a hybrid CPU?

YaSuenag replied:
Maybe yes. I checked `CPUID` 0Bh subleaf 1 on both Core Ultra 5 135U and 225U, they reports correct number of logical processors. However I cannot check all of hybrid CPUs and I'm not sure all of them should return correct value at here. Thus I'm not sure we can add `assert` here for hybrid CPU.

dholmes-ora commented on src/hotspot/share/jfr/metadata/metadata.xml:
```diff
@@ -852,7 +852,7 @@
 
   <Event name="CPUInformation" category="Operating System, Processor" label="CPU Information"
-         description="Characteristics and descriptions of the processor(s) the JVM is running on"
+         description="Characteristics and descriptions of the processor(s) the JVM is running on. Note that &quot;cores&quot; field would be incorrect on hybrid CPU."
```
Perhaps better stated on the cores field:
```
<Field type="uint" name="cores" label="Cores (approximation on a hybrid CPU" />
```
? But really this is for JFR folk to decide. Do these labels get used for GUI tools like JMC?

BTW I assume that "hwThreads" is the number of logical processors - right?

YaSuenag replied:
> BTW I assume that "hwThreads" is the number of logical processors - right?

Yes

YaSuenag replied:
> Perhaps better stated on the cores field:

Thanks for your comment! I added that sentence to `description` attribute in `Field` element.
After this change, flight record contains following metadata. I believe it would be parsed in some tools like JMC (JMC shows the description as a tooltip)

```
$ jfr metadata --events jdk.CPUInformation /tmp/test.jfr
@Name("jdk.CPUInformation")
@Category({"Operating System", "Processor"})
@Label("CPU Information")
@Description("Characteristics and descriptions of the processor(s) the JVM is running on")
class CPUInformation extends jdk.jfr.Event {
  @Label("Start Time")
  @Timestamp("TICKS")
  long startTime;

  @Label("Type")
  String cpu;

  @Label("Description")
  String description;

  @Unsigned
  @Label("Sockets")
  int sockets;

  @Unsigned
  @Label("Cores")
  @Description("Approximation on a hybrid CPU")
  int cores;

  @Unsigned
  @Label("Hardware Threads")
  int hwThreads;
}
```

YaSuenag replied:
@egahlin Can you approve this PR? Thanks.
